apps/chat/src/store/marketplace/marketplace.selectors.ts

```diff
@@ -107,6 +109,48 @@ const selectSourceTypes = createSelector(
   },
 );
 
+const selectToolsetSourceTypes = createSelector(
+  [ToolsetSelectors.selectToolsets],
+  (toolsets: ToolsetModel[]) => {
+    const sourceTypes = new Set<SourceType>([]);
+    if (toolsets.length === 0) {
+      return [];
+    }
+
+    let hasMyToolsets = false;
+    let hasPublic = false;
+    let hasSharedWithMe = false;
+
+    for (const toolset of toolsets) {
+      if (!hasMyToolsets && isMyToolset(toolset)) {
+        hasMyToolsets = true;
+        sourceTypes.add(SourceType.MyToolsets);
+      }
+
+      if (!hasPublic && isMarketplaceEntityPublic(toolset)) {
+        hasPublic = true;
+        sourceTypes.add(SourceType.Public);
+      }
+
+      if (
+        !hasSharedWithMe &&
+        !isMyToolset(toolset) &&
+        !isMarketplaceEntityPublic(toolset)
+      ) {
+        hasSharedWithMe = true;
+        sourceTypes.add(SourceType.SharedWithMe);
+      }
+
+      // Early exit optimization
+      if (hasMyToolsets && hasPublic && hasSharedWithMe) {
```

Comment:
check size instead of 3 boolean variables
```suggestion
      if (sourceTypes.size === 3) {
```

Reply:
Fixed as suggested.